[vcpkg/scripts] Add a way to get cmake compiler settings/flags (#12936)
microsoft/vcpkg: scripts/get_cmake_vars/CMakeLists.txt

    string(REGEX REPLACE "( |^)/" "\\1-" CMAKE_${flag}_FLAGS "${CMAKE_${flag}_FLAGS}")
    string(STRIP "${CMAKE_${flag}_FLAGS}" CMAKE_${flag}_FLAGS)
    string(APPEND OUTPUT_STRING "set(${VCPKG_VAR_PREFIX}_${flag}FLAGS \" ${CMAKE_${flag}_FLAGS}\")\n")
    foreach(config IN LISTS VCPKG_CONFIGS)
        string(REGEX REPLACE " +" " " CMAKE_${flag}_FLAGS_${config} "${CMAKE_${flag}_FLAGS_${config}}")
        string(REGEX REPLACE "( |^)/" "\\1-" CMAKE_${flag}_FLAGS_${config} "${CMAKE_${flag}_FLAGS_${config}}")
        string(STRIP "${CMAKE_${flag}_FLAGS_${config}}" CMAKE_${flag}_FLAGS_${config})
        string(APPEND OUTPUT_STRING "set(${VCPKG_VAR_PREFIX}_${flag}FLAGS_${config} \"${CMAKE_${flag}_FLAGS_${config}}\")\n")
        set(COMBINED_${flag}_${config}_FLAGS "${CMAKE_${flag}_FLAGS} ${CMAKE_${flag}_FLAGS_${config}}")
        string(REGEX REPLACE " +" " " COMBINED_${flag}_${config}_FLAGS "${COMBINED_${flag}_${config}_FLAGS}")
        string(STRIP "${COMBINED_${flag}_${config}_FLAGS}" COMBINED_${flag}_${config}_FLAGS)
        string(APPEND OUTPUT_STRING "set(${VCPKG_VAR_PREFIX}_COMBINED_${flag}FLAGS_${config} \"${COMBINED_${flag}_${config}_FLAGS}\")\n")

>     endforeach()

Reviewer: What's up with all this?

Author: - since cl nor the linker cares about `-` vs `/` I convert them directly to `-` which is a requirement for make based ports. 
- I remove any double spaces from the flags
- I strip trailing whitespaces
- Instead of manually combining flags later in the scripts I already export a combined flag here.

# foreach(prop IN LISTS add_props)
    # get_directory_property(${prop} ${prop})
    # string(APPEND OUTPUT_STRING "set(${VCPKG_VAR_PREFIX}_${prop} ${${prop}})\n")
    # foreach(config IN LISTS VCPKG_CONFIGS)
        # get_directory_property(${prop}_${config} ${prop}_${config})
        # string(APPEND OUTPUT_STRING "set(${VCPKG_VAR_PREFIX}_${prop}_${config} ${${prop}_${config}})\n")
    # endforeach()
# endforeach()
set(ENV_VARIABLES PATH INCLUDE C_INCLUDE_PATH CPLUS_INCLUDE_PATH LIB LIBPATH LIBRARY_PATH LD_LIBRARY_PATH )
foreach(_env IN LISTS ENV_VARIABLES)
    IF(WIN32)
        string(REPLACE "\\" "/" ENV_${_env} "$ENV{${_env}}")
        string(APPEND OUTPUT_STRING "set(${VCPKG_VAR_PREFIX}_ENV_${_env} \"${ENV_${_env}}\")\n")
    else()
        string(APPEND OUTPUT_STRING "set(${VCPKG_VAR_PREFIX}_ENV_${_env} \"$ENV{${_env}}\")\n")
    endif()
endforeach()

foreach(addvar IN LISTS VCPKG_ADDITIONAL_OUTPUT_VARS)
    string(REPLACE "CMAKE_" outvar "${addvar}")
    string(APPEND OUTPUT_STRING "set(${VCPKG_VAR_PREFIX}_${outvar} \"${${addvar}}\")\n")
endforeach()

file(WRITE "${VCPKG_OUTPUT_FILE}" "${OUTPUT_STRING}")

# Programs:
# CMAKE_AR
# CMAKE_<LANG>_COMPILER_AR (Wrapper)

> # CMAKE_RANLIB¶
> # CMAKE_<LANG>_COMPILER_RANLIB¶

Reviewer: ```suggestion
# CMAKE_RANLIB
# CMAKE_<LANG>_COMPILER_RANLIB
```

# CMAKE_STRIP
# CMAKE_NM 
# CMAKE_OBJDUMP
# CMAKE_DLLTOOL 
# CMAKE_MT
# CMAKE_LINKER

# CMAKE_C_COMPILER 
# CMAKE_CXX_COMPILER 

> # CMAKE_RC_COMPILER <--

Reviewer: ```suggestion
# CMAKE_RC_COMPILER
```

what's up with this arrow?

# CMAKE_RC_FLAGS

# Flags:
# CMAKE_<LANG>_FLAGS
# CMAKE_<LANG>_FLAGS_<CONFIG>
# CMAKE_SHARED_LINKER_FLAGS
# CMAKE_STATIC_LINKER_FLAGS

> # CMAKE_STATIC_LINKER_FLAGS_<CONFIG>.

Reviewer: ```suggestion
# CMAKE_STATIC_LINKER_FLAGS_<CONFIG>
```

# CMAKE_EXE_LINKER_FLAGS

> # CMAKE_EXE_LINKER_FLAGS_<CONFIG> (CMAKE_C_FLAGS_)

Reviewer: ```suggestion
# CMAKE_EXE_LINKER_FLAGS_<CONFIG>
```

list(REMOVE_DUPLICATES VCPKG_DEFAULT_ENV_VARS_TO_CHECK)

#Flags to check. Flags are a bit special since they are configuration aware. 
set(FLAGS ${VCPKG_LANGUAGES} RC SHARED_LINKER STATIC_LINKER EXE_LINKER)
foreach(flag IN LISTS FLAGS)
    list(APPEND VCPKG_DEFAULT_FLAGS_TO_CHECK CMAKE_${flag}_FLAGS)
endforeach()
list(REMOVE_DUPLICATES VCPKG_DEFAULT_FLAGS_TO_CHECK)

# TODO if ever necessary: Properties to check

set(VCPKG_VAR_PREFIX "VCPKG_DETECTED" CACHE STRING "Variable prefix to use for detected flags")
set(VCPKG_VARS_TO_CHECK "${VCPKG_DEFAULT_VARS_TO_CHECK}" CACHE STRING "Variables to check. If not given there is a list of defaults")
set(VCPKG_FLAGS_TO_CHECK "${VCPKG_DEFAULT_FLAGS_TO_CHECK}" CACHE STRING "Variables to check. If not given there is a list of defaults")
set(VCPKG_ENV_VARS_TO_CHECK "${VCPKG_DEFAULT_ENV_VARS_TO_CHECK}" CACHE STRING "Variables to check. If not given there is a list of defaults")

if(NOT VCPKG_OUTPUT_FILE)
    message(FATAL_ERROR "VCPKG_OUTPUT_FILE is required to be defined")
endif()

if(CMAKE_BUILD_TYPE)
    string(TOUPPER "${CMAKE_BUILD_TYPE}" VCPKG_CONFIGS)
else()

>     set(VCPKG_CONFIGS "DEBUG;RELEASE")

Reviewer: ```suggestion
    set(VCPKG_CONFIGS DEBUG RELEASE)
```

seems better?

    else()
        string(APPEND OUTPUT_STRING "set(${VCPKG_VAR_PREFIX}_ENV_${_env} \"$ENV{${_env}}\")\n")
    endif()
endforeach()

foreach(flag IN LISTS VCPKG_FLAGS_TO_CHECK)
    string(REGEX REPLACE " +" " " ${flag} "${${flag}}")
    string(REGEX REPLACE "( |^)/" "\\1-" ${flag} "${${flag}}")
    string(STRIP "${${flag}}" ${flag})
    string(APPEND OUTPUT_STRING "set(${VCPKG_VAR_PREFIX}_RAW_${flag} \" ${${flag}}\")\n")
    foreach(config IN LISTS VCPKG_CONFIGS)
        string(REGEX REPLACE " +" " " ${flag}_${config} "${${flag}_${config}}")
        string(REGEX REPLACE "( |^)/" "\\1-" ${flag}_${config} "${${flag}_${config}}")
        string(STRIP "${${flag}_${config}}" ${flag}_${config})
        string(APPEND OUTPUT_STRING "set(${VCPKG_VAR_PREFIX}_RAW_${flag}_${config} \"${CMAKE_${flag}_FLAGS_${config}}\")\n")
        set(COMBINED_${flag}_${config} "${${flag}} ${${flag}_${config}}")
        string(REGEX REPLACE " +" " " COMBINED_${flag}_${config} "${COMBINED_${flag}_${config}}")
        string(STRIP "${COMBINED_${flag}_${config}}" COMBINED_${flag}_${config})
        string(APPEND OUTPUT_STRING "set(${VCPKG_VAR_PREFIX}_${flag}_${config} \"${COMBINED_${flag}_${config}}\")\n")

>     endforeach()

Reviewer: Could we get a comment here describing what this regex and strip stuff does?

# Environment variables to check. 
list(APPEND VCPKG_DEFAULT_ENV_VARS_TO_CHECK PATH INCLUDE C_INCLUDE_PATH CPLUS_INCLUDE_PATH LIB LIBPATH LIBRARY_PATH LD_LIBRARY_PATH)
list(REMOVE_DUPLICATES VCPKG_DEFAULT_ENV_VARS_TO_CHECK)

#Flags to check. Flags are a bit special since they are configuration aware. 
set(FLAGS ${VCPKG_LANGUAGES} RC SHARED_LINKER STATIC_LINKER EXE_LINKER)
foreach(flag IN LISTS FLAGS)
    list(APPEND VCPKG_DEFAULT_FLAGS_TO_CHECK CMAKE_${flag}_FLAGS)
endforeach()
list(REMOVE_DUPLICATES VCPKG_DEFAULT_FLAGS_TO_CHECK)

# TODO if ever necessary: Properties to check

set(VCPKG_VAR_PREFIX "VCPKG_DETECTED" CACHE STRING "Variable prefix to use for detected flags")
set(VCPKG_VARS_TO_CHECK "${VCPKG_DEFAULT_VARS_TO_CHECK}" CACHE STRING "Variables to check. If not given there is a list of defaults")
set(VCPKG_FLAGS_TO_CHECK "${VCPKG_DEFAULT_FLAGS_TO_CHECK}" CACHE STRING "Variables to check. If not given there is a list of defaults")
set(VCPKG_ENV_VARS_TO_CHECK "${VCPKG_DEFAULT_ENV_VARS_TO_CHECK}" CACHE STRING "Variables to check. If not given there is a list of defaults")

if(NOT VCPKG_OUTPUT_FILE)
    message(FATAL_ERROR "VCPKG_OUTPUT_FILE is required to be defined")
endif()

> if(CMAKE_BUILD_TYPE)
>     string(TOUPPER "${CMAKE_BUILD_TYPE}" VCPKG_CONFIGS)
> else()
>     set(VCPKG_CONFIGS DEBUG RELEASE)
> endif()

Reviewer 2: We should never be calling this file without a build type
```suggestion
if(NOT CMAKE_BUILD_TYPE)
    message(FATAL_ERROR "CMAKE_BUILD_TYPE is required to be defined")
else()
    string(TOUPPER "${CMAKE_BUILD_TYPE}" VCPKG_CONFIGS)
endif()

```

project(get_cmake_vars LANGUAGES ${VCPKG_LANGUAGES})

> # These seem to be empty.
> # set(add_props INCLUDE_DIRECTORIES COMPILE_DEFINITIONS COMPILE_OPTIONS LINK_DIRECTORIES LINK_OPTIONS)
> # foreach(prop IN LISTS add_props)
>     # get_directory_property(${prop} ${prop})
>     # string(APPEND OUTPUT_STRING "set(${VCPKG_VAR_PREFIX}_${prop} ${${prop}})\n")
>     # foreach(config IN LISTS VCPKG_CONFIGS)
>         # get_directory_property(${prop}_${config} ${prop}_${config})
>         # string(APPEND OUTPUT_STRING "set(${VCPKG_VAR_PREFIX}_${prop}_${config} ${${prop}_${config}})\n")
>     # endforeach()
> # endforeach()
> 

Reviewer 2: ```suggestion
```

cmake_minimum_required(VERSION 3.17)

set(VCPKG_LANGUAGES "C;CXX" CACHE STRING "LANGUAGES TO ENABLE FOR THIS PROJECT")
set(VCPKG_VAR_PREFIX "VCPKG_DETECTED" CACHE STRING "Variable prefix to use for detected flags")
set(VCPKG_CONFIGS "DEBUG;RELEASE" CACHE STRING "Configurations to detect flags for")
set(VCPKG_OUTPUT_FILE "${CMAKE_BINARY_DIR}/../detected_flags.cmake.${VCPKG_TARGET_TRIPLET}.log" CACHE STRING "Output CMake files")
set(VCPKG_ADDITIONAL_OUTPUT_VARS "" CACHE STRING "Additional variables to output.")

set(OUTPUT_STRING)
project(get_cmake_vars LANGUAGES ${VCPKG_LANGUAGES})

if(MSVC)
    string(APPEND OUTPUT_STRING "set(${VCPKG_VAR_PREFIX}_MSVC TRUE)\n")
endif()
string(APPEND OUTPUT_STRING "set(${VCPKG_VAR_PREFIX}_CROSSCOMPILING ${CMAKE_CROSSCOMPILING})\n")
string(APPEND OUTPUT_STRING "set(${VCPKG_VAR_PREFIX}_SYSEM_NAME \"${CMAKE_SYSTEM_NAME}\")\n")
string(APPEND OUTPUT_STRING "set(${VCPKG_VAR_PREFIX}_SYSTEM_PROCESSOR \"${CMAKE_SYSTEM_PROCESSOR}\")\n")
# other:
# CMAKE_C_STANDARD

> # CMAKE_CXX_STANDARD

Author: Might need those in the future

Reviewer: Is there a reason that we aren't doing this now?

Author: needs a compile flag conversion. So it is probably just easier to add `/std:c++17` or similar directly in the toolchain and also setting the cmake standard variables. Especially meson seems a bit special on how to set the standard. 
```
cpp_std | none | none, c++98, c++03, c++11, c++14, c++17,c++1z, gnu++03, gnu++11, gnu++14, gnu++17, gnu++1z,vc++14, vc++17, vc++latest | C++ language standard to use
-- | -- | -- | --
```

Reviewer: Hehe, I think this might be partly my fault...

# foreach(prop IN LISTS add_props)
    # get_directory_property(${prop} ${prop})
    # string(APPEND OUTPUT_STRING "set(${VCPKG_VAR_PREFIX}_${prop} ${${prop}})\n")
    # foreach(config IN LISTS VCPKG_CONFIGS)
        # get_directory_property(${prop}_${config} ${prop}_${config})
        # string(APPEND OUTPUT_STRING "set(${VCPKG_VAR_PREFIX}_${prop}_${config} ${${prop}_${config}})\n")
    # endforeach()
# endforeach()
set(ENV_VARIABLES PATH INCLUDE C_INCLUDE_PATH CPLUS_INCLUDE_PATH LIB LIBPATH LIBRARY_PATH LD_LIBRARY_PATH )
foreach(_env IN LISTS ENV_VARIABLES)
    IF(WIN32)
        string(REPLACE "\\" "/" ENV_${_env} "$ENV{${_env}}")
        string(APPEND OUTPUT_STRING "set(${VCPKG_VAR_PREFIX}_ENV_${_env} \"${ENV_${_env}}\")\n")
    else()
        string(APPEND OUTPUT_STRING "set(${VCPKG_VAR_PREFIX}_ENV_${_env} \"$ENV{${_env}}\")\n")
    endif()
endforeach()

foreach(addvar IN LISTS VCPKG_ADDITIONAL_OUTPUT_VARS)
    string(REPLACE "CMAKE_" outvar "${addvar}")
    string(APPEND OUTPUT_STRING "set(${VCPKG_VAR_PREFIX}_${outvar} \"${${addvar}}\")\n")
endforeach()

file(WRITE "${VCPKG_OUTPUT_FILE}" "${OUTPUT_STRING}")

> # Not useful?
> # CMAKE_<LANG>_STANDARD_INCLUDE_DIRECTORIES
> # CMAKE_<LANG>_STANDARD_LIBRARIES
> 
> # CMAKE_<LANG>_IMPLICIT_LINK_FRAMEWORK_DIRECTORIES
> # CMAKE_<LANG>_IMPLICIT_INCLUDE_DIRECTORIES
> # CMAKE_<LANG>_IMPLICIT_LINK_DIRECTORIES
> # CMAKE_<LANG>_IMPLICIT_LINK_LIBRARIES
> 
> # CMAKE_LINK_INTERFACE_LIBRARIES -> LIBS

Author: ```suggestion
```

> 
> # Dir properties:
> # COMPILE_DEFINITIONS(_CONFIG)
> # COMPILE_OPTIONS(_CONFIG)
> # INCLUDE_DIRECTORIES
> # LINK_DIRECTORIES
> # LINK_OPTIONS
> 

Author: ```suggestion
```
